meta-facebook/yv4-sd/src/platform/plat_pldm_monitor.c

```diff
@@ -0,0 +1,251 @@
+		(struct pldm_set_state_effecter_states_req *)buf;
+	uint8_t *completion_code_p = resp;
+	*resp_len = 1;
+
+
+	if (req_p->composite_effecter_count < 0x01 || req_p->composite_effecter_count > 0x08) {
+		*completion_code_p = PLDM_ERROR_INVALID_DATA;
+		return PLDM_SUCCESS;
+	}
+
+
+	if (len != (PLDM_SET_STATE_EFFECTER_REQ_NO_STATE_FIELD_BYTES +
+		    sizeof(set_effecter_state_field_t) * req_p->composite_effecter_count)) {
+		*completion_code_p = PLDM_ERROR_INVALID_LENGTH;
+		return PLDM_SUCCESS;
+	}
+
+	struct plat_state_effecter_info *info_p = find_state_effecter_info(req_p->effecter_id);
+
+	if (!info_p) {
+		LOG_ERR("Can't find effecter ID (0x%x) info", req_p->effecter_id);
+		*completion_code_p = PLDM_ERROR_INVALID_DATA;
+		return PLDM_SUCCESS;
+	}
+
+	switch (info_p->entity_type) {
+	case PLDM_ENTITY_IO_CONTROLLER:
+		set_effecter_state_gpio_handler(buf, len, resp, resp_len,
+						(uint8_t)(info_p->effecter_id & GENMASK(7, 0)));
+		break;
+	case PLDM_ENTITY_SUB_CHASSIS:
+		plat_pldm_set_effecter_state_host_power_control(buf, len, resp, resp_len);
+		break;
+	default:
+		LOG_ERR("Unsupport entity type, (%d)", info_p->entity_type);
+		*completion_code_p = PLDM_ERROR_INVALID_DATA;
+		break;
+	}
+
+	return PLDM_SUCCESS;
+}
+
+uint8_t plat_pldm_get_state_effecter_state_handler(const uint8_t *buf, uint16_t len, uint8_t *resp,
+						   uint16_t *resp_len)
+{
+	CHECK_NULL_ARG_WITH_RETURN(buf, PLDM_ERROR);
+	CHECK_NULL_ARG_WITH_RETURN(resp, PLDM_ERROR);
+	CHECK_NULL_ARG_WITH_RETURN(resp_len, PLDM_ERROR);
+	CHECK_ARG_WITH_RETURN(!len, PLDM_ERROR);
+
+	struct pldm_get_state_effecter_states_req *req_p =
+		(struct pldm_get_state_effecter_states_req *)buf;
+	uint8_t *completion_code_p = resp;
+	*resp_len = 1;
+
+	struct plat_state_effecter_info *info_p = find_state_effecter_info(req_p->effecter_id);
+
+	if (!info_p) {
+		LOG_ERR("Can't find effecter ID (0x%x) info", req_p->effecter_id);
+		*completion_code_p = PLDM_ERROR_INVALID_DATA;
+		return PLDM_SUCCESS;
+	}
+
+	switch (info_p->entity_type) {
+	case PLDM_ENTITY_IO_CONTROLLER:
+		get_effecter_state_gpio_handler(buf, len, resp, resp_len,
+						(uint8_t)(info_p->effecter_id & GENMASK(7, 0)));
+		break;
+	default:
+		LOG_ERR("Unsupport entity type, (%d)", info_p->entity_type);
+		*completion_code_p = PLDM_ERROR_INVALID_DATA;
+		break;
+	}
+
+	return PLDM_SUCCESS;
+}
+
+void plat_pldm_assign_gpio_effecter_id()
+{
+	for (uint8_t i = 0; i <= PLDM_PLATFORM_OEM_AST1030_GPIO_PIN_NUM_NAX; i++) {
+		plat_state_effecter_table[i].effecter_id = (PLAT_EFFECTER_ID_GPIO_HIGH_BYTE | i);
+	}
+}
+
+uint8_t plat_pldm_host_button_sequence(const uint8_t *power_sequence, uint16_t pressing_interval)
+{
+	uint8_t sts_cnt = 3;
+	uint8_t retry = 3;
+	I2C_MSG i2c_msg;
+	i2c_msg.bus = I2C_BUS1;
+	i2c_msg.target_addr = 0x21; //server board cpld addr
+	i2c_msg.tx_len = 2;
+	i2c_msg.rx_len = 0;
+	i2c_msg.data[0] = 0x00;
+
+	for (int i = 0; i < sts_cnt; i++) {
+		i2c_msg.data[1] = power_sequence[i];
```

Comment:
I am wondering if we can move all the monitor logic to common code and just define bunch of get APIs for info like plat_power_sequence(), plat_get_processing_interval(), so all this code doesn't need to be duplicated for a new platform?
Do you think the generic orchestrator logic will change from one platform to another?

Reply:
Host power control handler will use the arguments like POWER_ON_BUTTON_MSEC which are unique in each platform. If moveing to common, 5 power control action will all create duplicated weak functions still implement differently in other platforms.